allennlp/data/dataset_readers/language_modeling.py

```diff
@@ -1,97 +1,160 @@
-    Reads a text file and converts it into a ``Dataset`` suitable for training a language model.
-
-    Note that there's one issue that needs to be fixed before this is actually usable for language
-    modeling - the way start and end tokens for sentences are handled is not correct; we need to
-    add a sentence splitter before this will be done right.
+    Reads a text file and converts it into a ``Dataset`` suitable for training
+    a language model.
 
     Parameters
     ----------
-    tokens_per_instance : ``int``, optional (default=``None``)
-        If this is ``None``, we will have each training instance be a single sentence.  If this is
-        not ``None``, we will instead take all sentences, including their start and stop tokens,
-        line them up, and split the tokens into groups of this number, for more efficient training
-        of the language model.
+    batch_size : ``int``, optional (default=``20``)
+        Batch size to use in language modeling.
+    truncated_bptt_size : ``int``, optional (default=``35``)
+        The sequence length to use for truncated backpropagation through time.
+    fuzz_truncated_bptt_size : ``bool``, optional (default=``True``)
+        If True, randomly perturb the truncated_bptt_size between batches.
+    bidirectional : ``bool``, optional (default=``False``)
+        If True, generate instances for bidirectional language modeling.
     tokenizer : ``Tokenizer``, optional (default=``WordTokenizer()``)
         We use this ``Tokenizer`` for the text.  See :class:`Tokenizer`.
     token_indexers : ``Dict[str, TokenIndexer]``, optional (default=``{"tokens": SingleIdTokenIndexer()}``)
-        We use this to define the input representation for the text.  See :class:`TokenIndexer`.
-        Note that the `output` representation will always be single token IDs - if you've specified
-        a ``SingleIdTokenIndexer`` here, we use the first one you specify.  Otherwise, we create
-        one with default parameters.
+        We use this to define the input representation for the text.
+        See :class:`TokenIndexer`.
+    start_tokens : ``List[str]``, optional (default=``None``)
+        These are prepended to each line read by the dataset reader.
+    end_tokens : ``List[str]``, optional (default=``["</S>"]``)
+        These are appended to each line read by the dataset reader.
     """
     def __init__(self,
-                 tokens_per_instance: int = None,
+                 batch_size: int = 20,
+                 truncated_bptt_size: int = 35,
+                 fuzz_truncated_bptt_size: bool = True,
+                 bidirectional: bool = False,
                  tokenizer: Tokenizer = None,
                  token_indexers: Dict[str, TokenIndexer] = None,
-                 lazy: bool = False) -> None:
-        super().__init__(lazy)
+                 start_tokens: List[str] = None,
+                 end_tokens: List[str] = ["</S>"]) -> None:
+        super().__init__(lazy=False)
```

Comment:
I think setting `lazy=False` defeats the purpose of using `fuzz_truncated_bppt_size=True` during training. The point of using random sequence lengths is to prevent batches ending with same tokens each epoch, however this will not happen if `lazy=False` since `_read` will only be called once. It probably makes more sense to let users set `lazy` themselves and raise an error if `fuzz_truncated_bppt_size=True` and `lazy=False`.